### PR TITLE
feat: Zephyr auxiliary

### PR DESCRIPTION
**Pull request by cody82**

### CODE REVIEW

File: docs/modules_usage/zephyr_auxiliary.rst [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        text: True
+      type: pykiso.lib.connectors.cc_process:CCProcess
+
+Remarks:
+
+- The test_directory contains the path of the directory that contains the testcase.yaml of the Zephyr test project.
+- The test_name contains the test name that is listed in the testcase.yaml.
+
+Then you can start the zephyr test from a test function or also setup/teardown, whatever you prefer:
+
```
Comment (sebastianpfischer) [inline docs/modules_usage/zephyr_auxiliary.rst:55]: Maybe also provide the link to the readthedocs API description (auxiliaries/...)

File: examples/zephyr.yaml [hunk 2]
```diff
@@ -0,0 +1,29 @@
+auxiliaries:
+  # Zephyr test auxiliary
+  zephyr_aux:
+    connectors:
+      com: twister
+    config:
+      # The location of the Zephyr test project
+      test_directory: ./test_zephyr/zephyr_test_project
+      # The name of the test in the test directory
+      test_name: testing.ztest
```
Comment (sebastianpfischer) [inline examples/zephyr.yaml:10]: Not sure we need that as parameter

File: src/pykiso/lib/auxiliaries/zephyr.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+        test_directory: Optional[str] = None,
+        test_name: Optional[str] = None,
+        wait_for_start: bool = True,
+        **kwargs: dict,
+    ) -> None:
+        """Initialize the auxiliary
+
+        :param twister_path: Path to the twister tool
+        :param test_directory: The directory to search for the Zephyr test project
+        :param testcase_name: The name of the Zephyr test
```
Comment (sebclrsn) [inline src/pykiso/lib/auxiliaries/zephyr.py:62]: This parameter doesn't exist anymore

File: src/pykiso/lib/auxiliaries/zephyr.py [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+        self, test_directory: Optional[str] = None, test_name: Optional[str] = None
+    ) -> None:
+        """Start the Zephyr test
+
+        :param test_directory: The directory to search for the Zephyr test project. Defaults to the test_directory from YAML.
+        :param testcase_name: The name of the Zephyr test. Defaults to the testcase_name from YAML.
+        """
+        test_directory = (
+            test_directory if test_directory is not None else self.test_directory
+        )
```
Comment (sebclrsn) [inline src/pykiso/lib/auxiliaries/zephyr.py:91]:
```suggestion
        test_directory = test_directory or self.test_directory
```

File: src/pykiso/lib/auxiliaries/zephyr.py [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+        test_name = test_name if test_name is not None else self.test_name
```
Comment (sebclrsn) [inline src/pykiso/lib/auxiliaries/zephyr.py:92]:
```suggestion
        test_name = test_name or self.test_name
```

File: src/pykiso/lib/auxiliaries/zephyr.py [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+    def _readline(self) -> Optional[str]:
+        """Read a line from the twister process
+
+        :return: The line as string or None if the process has finished
+        """
+        if not self.running:
+            return None
+
+        while True:
+            msg = self.channel.cc_receive(timeout=None)["msg"]
```
Comment (sebclrsn) [inline src/pykiso/lib/auxiliaries/zephyr.py:141]:
```suggestion
            msg = self.channel.cc_receive(timeout=None).get("msg")
```
Better be safe than sorry ;)

File: src/pykiso/lib/auxiliaries/zephyr.py [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+                if "exit" in msg:
+                    self.running = False
+                    return None
+
+            # Twister outputs many lines without a real message, remove those
+            test_output_pattern = "- OUTPUT:"
+            output = line.find(test_output_pattern)
+            if output != -1:
+                message = line[output + len(test_output_pattern) :]
+                if len(message) == 0 or message.isspace():
```
Comment (sebclrsn) [inline src/pykiso/lib/auxiliaries/zephyr.py:155]:
```suggestion
                if not message.strip():
```

File: src/pykiso/lib/auxiliaries/zephyr.py [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+        if not self.running:
+            raise ZephyrError("Twister was not started so can not wait.")
+
+        # Read the twister console for log output and wait for exit
+        while True:
+            line = self._readline()
+            if line is None:
+                break
+            # Look for relevant messages in the output
+            if "OUTPUT:  PASS - " in line:
```
Comment (sebclrsn) [inline src/pykiso/lib/auxiliaries/zephyr.py:174]: Same here you could create 2 class variables 
`TEST_RESULT_PATTERN = "OUTPUT:  "`
`TEST_PASSED_PATTERN = "PASS - "`

-> `if self.TEST_RESULT_PATTERN + self.TEST_PASSED_PATTERN in line:`

File: src/pykiso/lib/auxiliaries/zephyr.py [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+
+        # Get the result from the xunit file
+        result = self._parse_xunit(str(Path(self.outdir) / "twister_report.xml"))
+
+        return result
+
+    def _parse_xunit(self, file_path: str) -> TestResult:
+        """Parse xunit file
+
+        :return: The result of a single test case
```
Comment (sebclrsn) [inline src/pykiso/lib/auxiliaries/zephyr.py:187]: file_path param docstring missing

**Review by sebclrsn — CHANGES_REQUESTED** [review]

File: examples/test_zephyr/test_zephyr.py [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+    def setUp(self):
+        """Hook method from unittest in order to execute code before test case run.
+        In this case the default setUp method is overridden, allowing us to apply the
+        retry_test_case's decorator. The syntax super() access to the BasicTest and
+        we will run the default setUp()
+        """
+        super().setUp()
+
+    def test_run(self):
+        """In this case the default test_run method is overridden and
```
[the remark this answers is not in the document]
Comment (cody82) [reply examples/test_zephyr/test_zephyr.py:38]: Done

File: src/pykiso/lib/auxiliaries/zephyr.py (hunk 3 above)
Comment (cody82) [reply src/pykiso/lib/auxiliaries/zephyr.py:66]: Done

File: src/pykiso/lib/auxiliaries/zephyr.py [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+
+        self.running = True
+        if self.wait_for_start:
+            # Read twister output until test case was started
+            while True:
+                line = self._readline()
+                if line is None:
+                    return
+
+                if "OUTPUT: START - " in line:
```
[the remark this answers is not in the document]
Comment (cody82) [reply src/pykiso/lib/auxiliaries/zephyr.py]: Done

File: src/pykiso/lib/auxiliaries/zephyr.py (hunk 6 above)
Comment (cody82) [reply src/pykiso/lib/auxiliaries/zephyr.py]: Done

File: src/pykiso/lib/auxiliaries/zephyr.py (hunk 7 above)
Comment (cody82) [reply src/pykiso/lib/auxiliaries/zephyr.py]: Done

File: src/pykiso/lib/auxiliaries/zephyr.py (hunk 9 above)
Comment (cody82) [reply src/pykiso/lib/auxiliaries/zephyr.py:190]: Done

File: src/pykiso/lib/auxiliaries/zephyr.py (hunk 8 above)
Comment (cody82) [reply src/pykiso/lib/auxiliaries/zephyr.py]: Done

File: examples/test_zephyr/test_zephyr.py [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+        # Interact with the running device under test if needed...
+
+        # Wait for Zephyr test to finish and collect test result from the target.
+        # If an error occurs, it will raise an exception and finish the test.
+        result: TestResult = zephyr_aux.wait_test()
+        logging.info(f"Zephyr test has finished: {result}")
+        # The test case can decide how to handle test results from device under test.
+        self.assertEqual(result, TestResult.PASSED)
+
+    def tearDown(self):
```
[the remark this answers is not in the document]
Comment (cody82) [reply examples/test_zephyr/test_zephyr.py]: Done

File: examples/test_zephyr/test_zephyr.py [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+
+    -> suite_id : set to 1
+    -> case_id : set to 1
+    -> aux_list : test case test_run, setUp, and tearDown executed using
+    aux1 (see yaml configuration file)
+    -> test_ids: [optional] store the requirements into the report
+    -> tag: [optional] allows the run of subset of tests
+    """
+
+    def setUp(self):
```
[the remark this answers is not in the document]
Comment (cody82) [reply examples/test_zephyr/test_zephyr.py]: Done

**Comment by sebastianpfischer** [comment]
Discarded because approach may be abandoned

**Pull request closed without merging**